exec/java-exec/src/main/java/org/apache/drill/exec/util/ExecutableTasksLatch.java

```diff
@@ -0,0 +1,255 @@
+    this.executor = executor;
+    thread = Thread.currentThread();
+    count = new AtomicInteger();
+    executableTasks = new LinkedList<>();
+    this.testCountDownLatch = testCountDownLatch == null ? NoOpControlsInjector.LATCH : testCountDownLatch;
+  }
+
+  /**
+   * Waits for an earliest submitted task to complete and removes task from a collection of known tasks.
+   * @throws ExecutionException if task threw an Exception during execution
+   * @throws InterruptedException if wait is interrupted
+   */
+  public void take() throws ExecutionException, InterruptedException {
+    final ExecutableTask<C> task = executableTasks.peek();
+    Preconditions.checkState(task != null, "No tasks are scheduled for execution");
+    while (ExecutableTask.State.COMPLETING.compareTo(task.state.get()) > 0) {
+      if (Thread.interrupted()) {
+        throw new InterruptedException();
+      }
+      LockSupport.park();
+    }
+    executableTasks.remove();
+    while (ExecutableTask.State.COMPLETING.compareTo(task.state.get()) == 0) {
+      Thread.yield();
+    }
+
+    if (task.exception != null) {
+      throw task.exception;
+    }
+  }
+
+  /**
+   * @return immutable collection of submitted for execution tasks that are not yet taken from
+   * the {@linkplain ExecutableTasksLatch}
+   */
+  public Collection<ExecutableTask<C>> getExecutableTasks() {
+    return ImmutableList.copyOf(executableTasks);
+  }
+
+  /**
+   * submits a task for execution by {@linkplain ExecutorService}
+   * @param callable task to execute
+   * @return newly created instance of {@linkplain ExecutableTask}
+   */
+  public ExecutableTask<C> execute(C callable) {
+    ExecutableTasksLatch.ExecutableTask<C> task = new ExecutableTasksLatch.ExecutableTask<>(callable, this);
+    executor.execute(task);
+    count.getAndIncrement();
+    if (!executableTasks.add(task)) {
+      task.cancel(true);
+      throw new IllegalStateException();
+    }
+    return task;
+  }
+
+  /**
+   * Wait for completion of all tasks (cancelling them if cancel returns true)
+   * @param cancel BooleanSupplier that should return <tt>true</tt> when tasks cancellation is requested
+   */
+  public void await(BooleanSupplier cancel) {
+    Preconditions.checkState(Thread.currentThread() == thread,
+        String.format("%s can be awaited only from the %s thread", getClass().getSimpleName(), thread.getName()));
+    boolean cancelled = false;
+    while (count.get() > 0) {
+      if (!cancelled && cancel.getAsBoolean()) {
+        executableTasks.forEach(task -> task.cancel(true));
+        cancelled = true;
+      } else {
+        LockSupport.park();
+      }
+    }
+  }
+
+  /**
+   * Helper class to wrap {@linkplain Callable}{@literal <Void>} with cancellation and waiting for completion support
+   *
+   */
+  public static final class ExecutableTask<C extends Callable<Void>> implements Runnable {
```

Comment:
Can you elaborate what this class has that is not already available in `FutureTask`?

Reply:
Please see my comment for `ExecutableTasksLatch.await()`.